torch_xla/csrc/ops/index_op.cpp

```diff
@@ -0,0 +1,163 @@
+#include "torch_xla/csrc/ops/index_op.h"
+#include <ATen/ExpandUtils.h>
+#include <ATen/Functions.h>
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "torch_xla/csrc/aten_xla_bridge.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/arithmetic_ir_ops.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+#include "torch_xla/csrc/ops/ops.h"
+#include "torch_xla/csrc/xla_lower_util.h"
+
+namespace torch_xla {
+namespace {
+
+void CheckIndexTensorTypes(at::TensorList indices) {
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      auto& type = tensor.type();
+      auto scalar_type = type.scalarType();
+      if (scalar_type != at::kLong && scalar_type != at::kByte) {
+        XLA_ERROR() << "Tensors used as indices must be long or byte tensors, "
+                       "found scalar type: "
+                    << scalar_type;
+      }
+    }
+  }
+}
+
+// Expands byte tensors (masks) into the equivalent indexing by LongTensors.
+// This is a version of at::native::expandByteTensors with style adjustments.
+std::vector<at::Tensor> ExpandByteTensors(const at::Tensor& self,
+                                          at::TensorList indices) {
+  std::vector<at::Tensor> result;
+  for (auto& index : indices) {
+    if (index.type().scalarType() == at::kByte) {
+      // The sizes of the ByteTensor mask must match the sizes of the
+      // corresponding dimensions in self.
+      for (int64_t j = 0; j < index.dim(); j++) {
+        int64_t src_idx = result.size() + j;
+        XLA_CHECK_EQ(index.size(j), self.size(src_idx))
+            << "The shape of the mask " << index.sizes() << " at index " << j
+            << " does not match the shape of the indexed tensor "
+            << self.sizes() << " at index " << src_idx;
+      }
+      // Replace with nonzeros.
+      auto nonzero = index.nonzero();
+      for (int64_t j = 0; j < index.dim(); j++) {
+        result.emplace_back(nonzero.select(1, j));
+      }
+    } else {
+      result.emplace_back(index);
+    }
+  }
+  return result;
+}
+
+// Transposes the tensor and indices together so that all the non-null indices
+// index the first k dimensions of the tensor. Returns the transposed tensor and
+// the reordered indices. For example:
+//  TransposeToFront(tensor, {nullptr, a, nullptr, b})
+// returns
+//  tensor.permute([1, 3, 0, 2]), {a, b}
+//
+// This is a simplified version of at::native::transposeToFront which better
+// fits our requirements.
+CanonicalIndexInfo TransposeToFront(at::Tensor base, at::TensorList indices) {
+  std::vector<int64_t> dims;
+  std::vector<at::Tensor> transposed_indices;
+  size_t base_rank = base.dim();
+  dims.reserve(base_rank);
+  XLA_CHECK_LE(indices.size(), base_rank);
+  for (size_t i = 0; i < indices.size(); i++) {
```

Comment:
Care when we do loops like there on calls the compiler cannot see through, as it keep calling the bound API for every iteration:

https://gcc.godbolt.org/z/BG5HlI

Reply:
Realistically, we can only have maybe a dozen of indices.

Reply:
Sure, but the thing can keep creating shapes for some APIs.
The one above, is not issue. But things which end up calling into tensor methods, they might be.

Reply:
Ok, I don't understand the resolution here, neither the comment about shapes. If it's about dynamic models, the shape sizes will change, not the rank. We're not going to have rank 100k anytime soon.

Reply:
It's OK here. Mine was only a note ;)